cirq/ops/controlled_gate.py

```diff
@@ -26,38 +26,76 @@
 class ControlledGate(raw_types.Gate):
     """Augments existing gates with a control qubit."""
 
-    def __init__(self, sub_gate: raw_types.Gate) -> None:
+    def __init__(self, sub_gate: raw_types.Gate,
+                 control_qubits: List[raw_types.QubitId] = None,
+                 num_unspecified_control_qubits: int = None) -> None:
```

Comment:
The interface in #1388 is simpler (just the number of qubits). Can we adopt it here?

Reply:
This one specifically adds support for partial set of controls supplied initially. I followed the interface suggested by Craig so will wait for him to weigh in on this one.

Reply:
After some more thought, I think I agree with @viathor that it's better to just have the number of control qubits. The basic reason comes down to the fact that, with this interface, we have no nice way of controlled the order, of saying "the first control is this one, then the second control is not specified, then the third control qubit is that one". But adding that to the class would be very arduous. Speaking more algebraically, I think the thing that's bothering me is that when the class is nested you can't flatten it.

Also, the way the default behavior of the class interacts with specifying control qubits feels off. Like, the default number of unspecified controls varies depending on whether or not you specified explicit controls.

So I think if we want to add explicit controls to gates, we should do it with a separate class. Cut it down to just the number of controls.

Reply:
I have added logic to automatically flatten nested ControlledGates. The caveat there is that the qubit order for controlling is not preserved but I don't this that the qubit order matters once it's flattened. Even if I separate the partial qubit specification into a separate class, that class will have to deal with the same flattening issue and will probably inherit the same complexity as this.

About the default behavior of num_unspecified_control_qubits, I agree that the current logic is clumsy but I think that should be solvable if I give some more thought to it.

Reply:
It is extremely important that the qubit order is preserved. If I call `cirq.control(obj, c)`, I expect `c` to end up in a specific consistent place regardless of `obj`.

Reply:
I'm probably not aware of all the nuances here. When I am thinking of automatically flattened nested ControlledGates, I am thinking in case of circuits only. 

The flattened gate is basically n controls &ed together and controlling the sub_gate in the same Moment. In this final picture I see no concept of "qubit order" because all the controls would have been eventually specified to make an op/circuit. If the control qubits have values |q0, q1,..., qn>, regardless of what the qubit order was in the gate, the operation would have the same effect, right?

I don't know who will ever care about the internal qubit order for the gate before it's converted to either an Operation or a Circuit.

Reply:
As per our offline discussion, I have simplified the interface to accept control_qubits and num_controls. Also made the internal list respect the qubit ordering.